src/TaylorTrunc.jl

```diff
@@ -0,0 +1,278 @@
+        V = nothing
+        VT = nothing
+        D_tilda = 0
+        C_tilda = 0
+
+        if u
+            for i in 1:k
+                D_tilda = D_tilda + D(i, prob.b, opn,t,1)
+            end
+            N = D_tilda
+            D_tilda = sqrt(D_tilda)
+
+            u = isunitary(prob.A/opn)
+            rs = log2i(k+1)
+            VS1 = calc_vs2(k,rs,prob.b,t, opn, D_tilda)
+        else
+            for i in 1:k
+                D_tilda = D_tilda + D(i, prob.b, opn,t,2)
+            end
+            N = D_tilda
+            D_tilda = sqrt(D_tilda)
+            VT = calc_vt(CPType)
+            u = isunitary(prob.A/opn)
+            rs = k
+            VS1 = calc_vs2(k,prob.b,t, opn, D_tilda)
+        end
+        if !(T)
+            VS2 = VS1
+            if u
+                for i in 0:k
+                    C_tilda = C_tilda + C(i, prob.u0, opn, t,1)
+                end
+                C_tilda = sqrt(C_tilda)
+                VS1 = calc_vs1(k,rs,prob.u0,t, opn, C_tilda)
+            else
+                for i in 0:k
+                    C_tilda = C_tilda + C(i, prob.u0, opn, t,2)
+                end
+                C_tilda = sqrt(C_tilda)
+                VS1 = calc_vs1(k,prob.u0,t, opn, C_tilda)
+            end
+            N = sqrt(C_tilda^2 + D_tilda^2)
+            V = CPType[C_tilda/N D_tilda/N; D_tilda/N -1*C_tilda/N]
+            V = matblock(V)
+        end
+        new{CPType, u, typeof(VS1), L, Array{CPType,2}, typeof(V), typeof(VS1), typeof(VS2),typeof(VT)}(k, t, H, 2, rs, C_tilda, D_tilda, N, V, VS1, VS2,VT)
+    end
+end
+
+function calc_vs1(k::Int, x::Vector, t::Real, opn::Real, C_tilda::T) where T
+    CPType = eltype(x)
+    VS1 = rand(CPType,2^k,2^k)
+    VS1[:,1] = zero(VS1[:,1])
+    for j in 0:k
+        VS1[(2^k - 2^(k-j) + 1),1] = sqrt(C(j, x, opn, t,2))/C_tilda
+    end
+    VS1 = -1*qr(VS1).Q
+    return VS1
+end
+
+function calc_vs1(k::Int, rs::Int, x::Vector, t::Real, opn::Real, C_tilda::T) where T
+    CPType = eltype(x)
+    VS1 = rand(CPType,2^rs,2^rs)
+    for j in 0:k
+        VS1[j+1,1] = sqrt(C(j, x, opn, t,1))/C_tilda
+    end
+    VS1 = -1*qr(VS1).Q
+    return VS1
+end
+
+function calc_vs2(k::Int, x::Vector, t::Real, opn::Real, D_tilda::T) where T
+    D(m) = norm(prob.b)*(opn*t)^(m-1)*t/factorial(m)
+    VS2 = rand(CPType,2^k,2^k)
+    VS2[:,1] = zero(VS2[:,1])
+    for j in 0:k-1
+        VS2[(2^k - 2^(k-j) + 1),1] = sqrt(D(j+1, x, opn, t, 2))/D_tilda
+    end
+    VS2 = -1*qr(VS2).Q
+    return VS2
+end
+
+function calc_vs2(k::Int, rs::Int, x::Vector, t::Real, opn::Real, D_tilda::T) where T
+    VS2 = rand(CPType,2^rs,2^rs)
+    for j in 0:k - 1
+        VS2[j+1,1] = sqrt(D(j+1, x, opn, t, 1))/D_tilda
+    end
+    VS2[k+1,1] = 0;
+    VS2 = -1*qr(VS2).Q
+    return VS2
+end
+
+function unitary_decompose(H::Matrix)
+    if isunitary(H)
+        return H
+    end
+    Mu = H/opnorm(H)
+    nbit, = size(H)
+    nbit = log2i(nbit)
+    B1 = convert(typeof(H),1/2*(Mu + Mu'))
```

Comment:
Again, it would be shorter if you use a type parameter in the function head here.